Add Layouts Submenu to Application Menu

rgeraskin opened the pull request:
## Add Layouts Submenu to Application Menu

### Description
This PR introduces a new layouts submenu in the application menu, providing users with quick access to layout options directly from the menu bar.

Closes: https://github.com/ianyh/Amethyst/issues/1080

### Changes Made
- **AppDelegate.swift**: Added 87 lines of code to implement the layouts submenu functionality
- **MainMenu.xib**: Updated the interface builder file to include the new submenu structure

### What's New
- ✨ New layouts submenu in the application menu
- 🎯 Enhanced user experience with easier layout access
- 📱 Improved menu navigation for layout management

### Files Modified
- `Amethyst/AppDelegate.swift` (+87 lines)
- `Amethyst/Base.lproj/MainMenu.xib` (+7 lines)

rgeraskin committed:
feat(menu): add layouts submenu

rgeraskin commented on Amethyst/AppDelegate.swift:
```diff
@@ -173,6 +184,66 @@ class AppDelegate: NSObject, NSApplicationDelegate {
+        // If no focused screen manager, fallback to the first screen manager
+        if focusedScreenManager == nil, let firstScreenManager = windowManager?.screenManager(at: 0) {
+            currentLayoutKey = firstScreenManager.currentLayout?.layoutKey
+        }
```
not sure that this is the correct way to find ScreenManager

ianyh replied:
I would treat it as an error if `focusedScreenManager` is nil. Maybe put a placeholder item in the submenu that just says it can't be found.

rgeraskin replied:
Fixed. Improved screen manager detection with a fallback chain:
1. Try `focusedScreenManager()` first
2. If nil (e.g., when clicking menu bar steals focus), detect the screen under the mouse cursor using `NSEvent.mouseLocation`
3. Fall back to `screenManager(at: 0)` as last resort
4. Show "Unable to determine current screen" placeholder only if all methods fail

This ensures the correct screen's layout is shown when clicking the menu bar on multi-monitor setups.

rgeraskin commented:
rebased on development branch

ianyh commented on Amethyst/AppDelegate.swift:
```diff
@@ -93,6 +99,11 @@ class AppDelegate: NSObject, NSApplicationDelegate {
         startAtLoginMenuItem?.state = (LoginServiceKit.isExistLoginItems(at: Bundle.main.bundlePath) ? .on : .off)
+
+        // Set up layouts menu delegate to refresh when opened
+        layoutsMenuItem?.submenu?.delegate = self
```
It might be better to do this when the main menu is opened, so that the list is available before the submenu is accessed.

rgeraskin replied:
Fixed. Changed the delegate from `layoutsMenuItem?.submenu` to `statusItemMenu`. The layouts menu is now populated in `menuWillOpen` when the main status item menu opens, making the list available before the user hovers over the Layouts submenu.

ianyh commented on Amethyst/AppDelegate.swift:
```diff
@@ -173,6 +184,66 @@ class AppDelegate: NSObject, NSApplicationDelegate {
+            let noLayoutsItem = NSMenuItem(title: "No layouts enabled", action: nil, keyEquivalent: "")
+            noLayoutsItem.isEnabled = false
+            submenu.addItem(noLayoutsItem)
+        }
```
This is minor, but can you just do this check before the loop and then return if it's empty?

rgeraskin replied:
Fixed. Moved the empty layouts check to the beginning of `populateLayoutsMenu()` with an early return.

ianyh commented on Amethyst/AppDelegate.swift:
```diff
@@ -173,6 +184,66 @@ class AppDelegate: NSObject, NSApplicationDelegate {
+            currentLayoutKey = firstScreenManager.currentLayout?.layoutKey
+        }
+
+        // Filter to only enabled layouts and add menu items
```
This logic isn't quite correct because it is technically possible for there to be multiple instances of the same layout in a space.

rgeraskin replied:
Fixed. Added `currentLayoutIndexValue` property to ScreenManager to expose the current layout index. The menu now compares by index instead of by layoutKey:

- Before: `currentLayout?.layoutKey == layoutKey` (would checkmark ALL instances of same layout type)
- After: `index == currentLayoutIndex` (only checkmarks the specific instance)

This correctly handles configurations like [Tall, Wide, Tall] where only the active instance shows a checkmark.

rgeraskin committed:
fix(menu): address PR review feedback for layouts submenu

- Populate layouts menu when main menu opens (not submenu)
- Add screen detection fallback using mouse cursor position
- Show error placeholder only when no screen manager available
- Move empty layouts check before loop with early return
- Fix multiple layout instances by comparing index instead of key
- Add currentLayoutIndexValue property to ScreenManager

ianyh commented on Amethyst/AppDelegate.swift:
```diff
@@ -173,6 +177,83 @@ class AppDelegate: NSObject, NSApplicationDelegate {
+                let amScreen = AMScreen(screen: nsScreen)
+                return windowManager?.screenManager(for: amScreen)
+            }
+            return windowManager?.screenManager(at: 0)
```
I would just return `nil` here.

rgeraskin replied:
Removed the screenManager(at: 0) fallback. Now returns nil if both focusedScreenManager() and mouse cursor screen detection fail, which will show the error placeholder.

ianyh reviewed (changes requested):
My point about the layout instances is that we probably want this to be derived from the actual list of layouts in the screen manager.

rgeraskin committed:
fix(menu): derive layouts from screen manager instead of global config

- Add layoutsInfo property to ScreenManager to expose actual layouts
- Remove screenManager(at: 0) fallback, return nil instead
- Use screen manager's layouts for menu instead of UserConfiguration
- Ensures indices match correctly for current layout detection

rgeraskin commented:
> My point about the layout instances is that we probably want this to be derived from the actual list of layouts in the screen manager.

Refactored to derive the menu from the screen manager's actual layouts instead of the global UserConfiguration.shared.layoutKeys():

- Added layoutsInfo property to ScreenManager that exposes the layouts array as [(key: String, name: String)]
- populateLayoutsMenu() now iterates over screenManager.layoutsInfo instead of global config
- This ensures indices match correctly since both menu items and currentLayoutIndex come from the same source

```
// Before: used global config (indices might not match screen manager)
let enabledLayoutKeys = UserConfiguration.shared.layoutKeys()
for (index, layoutKey) in enabledLayoutKeys.enumerated() { ... }
```

```
// After: derived from screen manager's actual layouts
let layouts = screenManager.layoutsInfo
for (index, layoutInfo) in layouts.enumerated() { ... }
```

ianyh commented on Amethyst/Managers/ScreenManager.swift:
```diff
@@ -56,6 +56,16 @@ final class ScreenManager<Delegate: ScreenManagerDelegate>: NSObject, Codable {
+    /// Returns layout info (key and name) for all layouts in this screen manager
+    var layoutsInfo: [(key: String, name: String)] {
+        return layouts.map { (key: $0.layoutKey, name: $0.layoutName) }
+    }
```
* Could you change layoutsInfo to use a defined struct type instead of the named tuple?
* Could you include whether or not it is the selected layout in that struct instead of exposing the index separately above?

rgeraskin replied:
done

ianyh reviewed (changes requested):
One comment, but otherwise I think it looks good.

rgeraskin committed:
refactor(menu): use LayoutMenuItemInfo struct instead of tuple

- Add LayoutMenuItemInfo struct with key, name, and isSelected properties
- Remove currentLayoutIndexValue property from ScreenManager
- Encapsulate selection logic in layoutsInfo property
- Simplify AppDelegate menu population loop
